### PR TITLE
tezos_rpc: library and /inject-operations

### PR DESCRIPTION
**Pull request by EduardoRFS**
## Depends

- [x] #445

## Problem

As mentioned at #429 ideally we should move away from Taquito, for this one of the requirements is that we need to be able to inject operations on Tezos.

## Solution

This PR exposes `Tezos_rpc.Inject_operations.execute` to actually inject the operations on the chain, under the hood it uses #429 work to make the HTTP requests to a Tezos node.

This also defines the `tezos_rpc` library to be extended in the future.

### CODE REVIEW

File: src/tezos/operation.mli
Comment (gsebil08) [inline src/tezos/operation.mli:5]: [OCaml newbie question] why is this type defined twice (in .ml and .mli)? I find this pretty error-prone (like modifying in .ml and not here)

File: src/tezos/operation.ml
Comment (gsebil08) [inline src/tezos/operation.ml:9]: At this point, I am not sure about what is _really_ the entrypoint (I think it is _the smart contract entrypoint_, am I right?), but should we go for a custom type instead of `string`?
I think going for a custom type could help us to get some domain specific: What is this string, functionally? => Can this string be empty? Can this string be only a space character? Does it have to match a given regex?

File: src/tezos/michelson.ml
Comment (gsebil08) [inline src/tezos/michelson.ml:13]: Is there a point to define this function globally while it is used only once in `operation:encoding`?

File: tests/test_tezos_interop.re
Comment (gsebil08) [inline tests/test_tezos_interop.re:635]: Why do we keep these lines? If they are important they should not be comments but documentation.
If they are not important and it is just "a way to retrieve them if needed" this is the job of GitHub, what do you think?

File: tests/test_tezos_interop.re
Comment (gsebil08) [inline tests/test_tezos_interop.re:579]: What about a clear convention for tests, something like 
>When calling _{function}_, successfully _{expected action}_

Because with the actual description, I don't know what is the scope of this test. Maybe the case `when calling forging with bad parameters, fails with this error` is not covered...

File: src/tezos_rpc/http.ml
Comment (gsebil08) [inline src/tezos_rpc/http.ml:4]: I don't see any test module, is it normal?

File: src/tezos_rpc/http.mli
Comment (gsebil08) [inline src/tezos_rpc/http.mli:1]: I feel like only the `http_post` is used, then should we already define it?

File: src/tezos_rpc/http.ml
Comment (gsebil08) [inline src/tezos_rpc/http.ml:25]:
```suggestion
  let%await body = http_request uri ~method_ data in
```

Comment (gsebil08) [reply src/tezos_rpc/http.ml:25]: If I understand correctly, of course 😅

**Review by gsebil08 — COMMENTED** [review]
First real review, hope my comments can be useful...

**Review by Zett98 — APPROVED** [review]

File: src/tezos/operation.mli
Comment (EduardoRFS) [reply src/tezos/operation.mli]: The mli is an interface for the ml(the implementation), so if you change one but not the other you're gonna get an error, think about mli in a similar fashion to a C header but checked against the implementation.

File: src/tezos/operation.ml
Comment (EduardoRFS) [reply src/tezos/operation.ml]: Yes, this is the contract entrypoint. About the functionality, it can be any string if I remember correctly up to 61 chars.

Your argument about enforcing it on the typer makes sense, just that it's not a considerable win here, but an considerable amount of complexity, especially because this entrypoint is likely gonna be a hard coded string.

I would say that this should not block this PR, but if you want to open another PR adding an `Entrypoint.t` feel free.

File: src/tezos/michelson.ml
Comment (EduardoRFS) [reply src/tezos/michelson.ml]: Important thing, it's not "globally", it's inside of a module, which is inside of a library, while it is accessible globally it's not like it's polluting the scope. It's quite different from Haskell in this sense.

But this was on the other PR and the reasons why I added on Michelson instead of just doing it inline, is because I was following a similar style from Tezos and thinking now(after the fact, which is biased), it makes sense conceptually, I would prefer to avoid comparing Michelson outside of this module in general.

File: tests/test_tezos_interop.re
Comment (EduardoRFS) [reply tests/test_tezos_interop.re]: That was a mistake of mine, I forgot it during the development, but noticed before merging, it was on the previous PR tho.

File: tests/test_tezos_interop.re
Comment (EduardoRFS) [reply tests/test_tezos_interop.re]: Please open an issue for this, but those tests are mostly regression test, they're not trying to enforce failure properties, just to ensure that the cases that we actually care are covered here.

File: src/tezos_rpc/http.ml
Comment (EduardoRFS) [reply src/tezos_rpc/http.ml:4]: Currently we don't have a lot of automated tests, and we're in crunch mode, but @d4hines is working on having more test infrastructure.

File: src/tezos_rpc/http.mli
Comment (EduardoRFS) [reply src/tezos_rpc/http.mli:1]: It's gonna be used in a future PR(which is now already opened), also it's a 2 lines implementation, so it's quite reasonable.

File: src/tezos_rpc/http.ml
Comment (EduardoRFS) [reply src/tezos_rpc/http.ml:25]: This typed, as you can see from the CI. But also nope, this implementation is correct.

https://github.com/marigold-dev/deku/pull/444/files#diff-698e4c768874ed323e94fa37737b6039c3ef8f7a6f5222ecd43cfaab42d59c3bR7

**Review by anchpop — APPROVED** [review]

File: src/tezos_rpc/http.ml
Comment (gsebil08) [reply src/tezos_rpc/http.ml:25]: Thanks, I guess I had something weird with my environment, since it showed me `unused variable`. Working great now. Thanks!

File: src/tezos_rpc/http.ml
Comment (gsebil08) [reply src/tezos_rpc/http.ml:4]: What about basic testing?
For instance, even if we can suppose that `yojson` is working, having tests like:
> when calling to_yojson on this object, it produces the following JSON

or also

> when calling to_yojson then from_yojson, it gives back the original object

If we always say "we can do that later" we will never do it... Moreover, some tests are just quick-win and ensure our domain is safe in the real world.

For instance, I think the function `http_request` isn't so difficult to test, am I wrong?
Or maybe we can start by only testing `http_post` function, which is exposed and used, and test the others later.
What do you think?

File: src/tezos/operation.ml
Comment (gsebil08) [reply src/tezos/operation.ml]: Opened #452

File: src/tezos/michelson.ml
Comment (gsebil08) [reply src/tezos/michelson.ml]: Didn't see it was inside a module. Thanks.

File: tests/test_tezos_interop.re
Comment (gsebil08) [reply tests/test_tezos_interop.re]: Created #453

File: src/tezos_rpc/http.ml
Comment (gsebil08) [reply src/tezos_rpc/http.ml:4]: Opened #454

**Pull request merged**
